fixes #11193, #11209 - Widget improvements and add status indicator

stbenjam opened the pull request:
Just something I noticed working with the plugin. This probably requires removing a column, so I picked IP address.  Any thoughts?

7 wide: http://i.imgur.com/tJhRCqZ.png
6 wide: http://i.imgur.com/l4IXpaa.png

The latter looks better to me!

stbenjam committed:
fixes #11193, #11209 - Widget improvements and add status indicator

lzap commented:
Ok works great, the only issue is N+1 which is introduced now. Can you remove it please?

```
2015-07-23T10:58:24 [app] [W] user: lzap
 | http://localhost:3000/
 | Unused Eager Loading detected
 |   Host::Discovered => [:interfaces]
 |   Remove from your finder: :includes => [:interfaces]
 | 
```

ohadlevy commented:
afair interface problem was already there :)

and, i would consider dropping the hostname too, I'm not sure if its
valuable?

lzap commented:
> afair interface problem was already there :)

Believe it or not, we have fixed it already :-)

## 

Later,
 Lukas #lzap Zapletal

stbenjam commented:
Ah, I did cause it. Interfaces include isn't required because I removed the IP.

I think 'name' is good, it's the main identifier of the host.  Unless you'd prefer to use the space to include some additional fact?

lzap commented:
It looks good. To be honest, my impression is that the most valuable would be just the MAC address. It's the only thing one can surely identify server with and server vendors _usually_ put stickers with MAC addresses on it. So instead hostname IMHO just pure MAC address could do it too. Optionally with vendor_id if present (the fact).

I was wondering if we can also review this line in this patch:

```
<h4 class="ca"><%= _('Discovered Host Pool') %> - <%= @current_permission %></h4
```

The `@current_permission` is always nil and the header does not look nice. This is here from the very beginning, I think this was a debugging leftover or something. I'd leave just the title, what you think?

ohadlevy commented:
1. since the hostname is customize and by default already uses mac, I think
   its fine to leave it as is, plus it allows the user to change the column as
   she likes.
2. I agree regarding the header, we could add the number of discovered
   hosts for example in parenthesis and cut the total lines from below.

lzap commented:
@stbenjam ok so please solve point (2) and we are good to go! Thanks for this.

stbenjam commented:
Ok, updated.   Thanks!
- made the header same left-aligned as the rest, and added total there
- added one more discovered host
- sort descending (show newest)

![overview](https://cloud.githubusercontent.com/assets/429763/8859833/d4b86b6a-314d-11e5-93d1-209c20ec8ce9.png)

stbenjam commented:
Welllll, while we're here tweaking the widget, how about this? 
- New in last 24 hours
- Reported in the last 7 days
- Haven't heard from the host in 7 days

![dash](https://cloud.githubusercontent.com/assets/429763/8861102/4826604a-3156-11e5-8330-238ed75d67b1.png)

lzap commented on app/views/dashboard/_discovery_widget_host_list.html.erb:
```diff
@@ -1,16 +1,16 @@
   <tbody>
   <% hosts.each do |host| %>
     <tr>
-      <%= render :partial => "discovered_hosts/discovered_host", :locals => {
```
I think we can delete this whole file completely. It is not used anymore. I mean the `discovered_hosts/discovered_host` partial.

stbenjam replied:
It's used by the Discovered Host index page

lzap commented on app/views/dashboard/_discovery_widget_host.html.erb:
```diff
@@ -0,0 +1,23 @@
+  <span class="glyphicon <%= status_glyph %>"
+      style="color: <%= status_color %>"
+      title="<%= status_message %>"/>
+</td>
```
I was wondering if we can put the icon in the same table cell as the name. Or maybe making its own cell not to enlarge when there is plenty of space. In extreme cases (very wide monitors) icons can be little bit far away from the name which is IMHO an UX issue and does not look nice.

lzap commented:
Ok I really like the idea, can we extract the icon code into a helper? I am going to use the same on the main discovery index page. Or if you want to stretchen the patch once again. Maybe rename the commit message a bit in this case because it went a little bit too far, but it's great addition!

I struggle myself to find the latest discovered host and this sorting with those icons make _a lot_ of sense! Thanks :-D

stbenjam commented:
Ok, done, let me know what you think

lzap commented on app/views/dashboard/_discovery_widget.html.erb:
```diff
@@ -1,10 +1,12 @@
-<% all_hosts = Host::Discovered.includes(:model, :interfaces).includes(:discovery_attribute_set).scoped %>
-<% hosts = all_hosts.limit(5) %>
+<% all_hosts = Host::Discovered.includes(:model).includes(:discovery_attribute_set).scoped %>
+<% hosts = all_hosts.order('created_at DESC').limit(6) %>
```
Can you implement the same ordering for the main discovered hosts index page as well? That would be useful. To be honest, I think we are currently unsorted.

lzap commented on app/views/dashboard/_discovery_widget_host.html.erb:
```diff
@@ -0,0 +1,4 @@
+<td><%= discovery_status_icon(host) %>&nbsp;<%= link_to trunc_with_tooltip(h(host.name)), discovered_host_path(host) %></td>
```
Was wondering if we can do a smaller blank space. It looks like `&nbsp;` is maybe too much. Maybe to add just few pixels to the `span` that holds the icon. Smaller the better.

stbenjam replied:
Not sure how to do that, can you give me an example? The `&nbsp;` prevents the icon from getting separated from the hostname, is there a better way to do that?

lzap commented:
Looks good, two nitpicks...

lzap commented on app/helpers/discovered_hosts_helper.rb:
```diff
@@ -44,4 +44,23 @@ def discovery_attribute(host, attr, default_value = _('N/A'))
+      status_color = '#4572A7'
+    end
+
+    "<span class='glyphicon #{status_glyph}' style='color: #{status_color}'
```
Was thinking like:

```
"<span class='glyphicon #{status_glyph}' style='color: #{status_color};padding-left:10px '
```

But I realized this won't work. In that case use `&thinsp;` instead which looks much better.

stbenjam replied:
`&thinsp;` is still a breaking space, so I used http://www.fileformat.info/info/unicode/char/202f/index.htm instead.  That seems to work better on the index page so the icon stays with the name.

lzap commented on app/models/host/discovered.rb:
```diff
@@ -25,6 +26,10 @@ class Host::Discovered < ::Host::Base
 
+  default_scope lambda {
+    where(taxonomy_conditions).order("hosts.created_at DESC")
+  }
```
What is this used for?

lzap replied:
I understand, this is from Host::Base. I wonder if there is a way to chain default scopes... I guess its not possible.

lzap commented on app/views/dashboard/_discovery_widget_host_list.html.erb:
```diff
@@ -2,15 +2,14 @@
   <% hosts.each do |host| %>
     <tr>
-      <%= render :partial => "discovered_hosts/discovered_host", :locals => {
-        :host => host, :hide_disk => true } %>
```
This was unused, right?

lzap commented on app/views/discovered_hosts/_discovered_host.html.erb:
```diff
@@ -1,12 +1,10 @@
 <% Setting::Discovered.discovery_fact_column_array.each do |fact_column| %>
 <td class="hidden-tablet hidden-xs"><%= host.facts_hash[fact_column.strip] || 'N/A' %></td>
 <% end %>
-<% end %>
```
Oh I see.

lzap commented:
Merged as d498e7485b43fba711880cc73fdb916f49c8ccdd, thank you!

lzap commented:
Actually, I have just realized, that last_report is not being updated because discovered host only check-in once, and then when a fact changes (hardware change). We should change this behavior and let them to check in regularly. More details in: http://projects.theforeman.org/issues/12310
